cdc_kafka/tracked_tables.py

```diff
@@ -0,0 +1,318 @@
+            latest_instance = sorted(capture_instances, key=lambda x: x['regex_matched_group'])[-1]
+        else:
+            raise Exception(f'Capture instance version strategy "{capture_instance_version_strategy}" not recognized.')
+        result.append(latest_instance['capture_instance'])
+
+    return result
+
+
+def build_tracked_tables_from_cdc_metadata(
+        db_conn: sqlalchemy.engine.Connection, topic_name_template: str, table_whitelist_regex: str,
+        table_blacklist_regex: str, snapshot_table_whitelist_regex: str, snapshot_table_blacklist_regex: str,
+        capture_instance_version_strategy: str, capture_instance_version_regex: str) -> List[TrackedTable]:
+    result = []
+
+    latest_names = get_latest_capture_instance_names(db_conn, capture_instance_version_strategy,
+                                                     capture_instance_version_regex)
+    logger.debug('Latest capture instance names determined by "%s" strategy: %s', capture_instance_version_strategy,
+                 sorted(latest_names))
+
+    cdc_metadata_rows = db_conn.execute(text(constants.CDC_METADATA_QUERY),
+                                        {'latest_capture_instance_names': latest_names})
+
+    snapshot_table_whitelist_regex = snapshot_table_whitelist_regex and re.compile(snapshot_table_whitelist_regex)
+    snapshot_table_blacklist_regex = snapshot_table_blacklist_regex and re.compile(snapshot_table_blacklist_regex)
+
+    for cdc_metadata_row in cdc_metadata_rows:
+        fq_table_name = f'{cdc_metadata_row["schema_name"]}.{cdc_metadata_row["table_name"]}'
+
+        can_snapshot = False
+
+        if table_whitelist_regex and not re.match(table_whitelist_regex, fq_table_name):
+            logger.debug('Table %s excluded by whitelist', fq_table_name)
+            continue
+
+        if table_blacklist_regex and re.match(table_blacklist_regex, fq_table_name):
+            logger.debug('Table %s excluded by blacklist', fq_table_name)
+            continue
+
+        if snapshot_table_whitelist_regex and re.match(snapshot_table_whitelist_regex, fq_table_name):
+            logger.debug('Table %s WILL be snapshotted due to whitelisting', fq_table_name)
+            can_snapshot = True
+
+        if snapshot_table_blacklist_regex and re.match(snapshot_table_whitelist_regex, fq_table_name):
+            logger.debug('Table %s will NOT be snapshotted due to blacklisting', fq_table_name)
+            can_snapshot = False
+
+        # ORDER BY in the SQL query ensures this means we're on to a new table:
+        if cdc_metadata_row['change_table_ordinal'] == 1:
+            topic_name = topic_name_template.format(**cdc_metadata_row)
+            tracked_table = TrackedTable(
+                db_conn, cdc_metadata_row['schema_name'], cdc_metadata_row['table_name'],
+                cdc_metadata_row['capture_instance_name'], topic_name, can_snapshot)
+            result.append(tracked_table)
+
+        tracked_table.add_field(TrackedField(
```

Comment:
it appears tracked_table can never be None here because the first cdc_metadata_row is always the the first field ?  Is that right?  This definitely made me pause and think for a while

Reply:
That is correct. I will add comments to make that clearer, thanks.